pysmt/solvers/msat.py

```diff
@@ -67,31 +63,79 @@ def __init__(self, environment, logic, options=None, debugFile=None):
         self.converter = MSatConverter(environment, self.msat_env)
         return
 
+
+    def _prepare_config(self, options, debugFile=None):
+        """Sets the relevant options in self.msat_config"""
+        if options.generate_models:
+            check = mathsat.msat_set_option(self.msat_config, "model_generation",
+                                            "true")
+            assert check == 0
+
+        if options.unsat_cores_mode is not None:
+            check = mathsat.msat_set_option(self.msat_config,
+                                            "unsat_core_generation",
+                                            "1")
+            assert check == 0
+
+
+        if debugFile is not None:
+            mathsat.msat_set_option(self.msat_config,
+                                    "debug.api_call_trace", "1")
+            mathsat.msat_set_option(self.msat_config,
+                                    "debug.api_call_trace_filename",
+                                    debugFile)
+
+
     @clear_pending_pop
-    def reset_assertions(self):
+    def _reset_assertions(self):
         mathsat.msat_reset_env(self.msat_env)
-        return
 
     @clear_pending_pop
     def declare_variable(self, var):
         self.converter.declare_variable(var)
 
     @clear_pending_pop
-    def add_assertion(self, formula, named=None):
+    def _add_assertion(self, formula, named=None):
         self._assert_is_boolean(formula)
-        term = self.converter.convert(formula)
 
+        result = formula
+        if self.options.unsat_cores_mode == "named":
+            # If we want named unsat cores, we need to rewrite the
+            # formulae as implications
+            key = self.mgr.FreshSymbol(template="_assertion_%d")
+            result = (key, named, formula)
+            formula = self.mgr.Implies(key, formula)
+
+        term = self.converter.convert(formula)
         res = mathsat.msat_assert_formula(self.msat_env, term)
 
         if res != 0:
             msat_msg = mathsat.msat_last_error_message(self.msat_env)
             raise InternalSolverError(msat_msg)
-        return
 
+        return result
+
+    def _named_assertions(self):
+        if self.options.unsat_cores_mode == "named":
+            return [t[0] for t in self.assertions]
+        return None
+
+    def _named_assertions_map(self):
+        if self.options.unsat_cores_mode == "named":
+            return {t[0] : (t[1],t[2]) for t in self.assertions}
+        return None
 
     @clear_pending_pop
-    def solve(self, assumptions=None):
+    def _solve(self, assumptions=None):
         res = None
+
+        n_ass = self._named_assertions()
```

Comment:
Doing this at solving time seems unnecessarily complicated. Why can't we just assert the assumption when calling add_assertion (see comment above)?